lib/vm/src/mmap.rs

```diff
@@ -256,6 +314,186 @@ impl Mmap {
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    /// Copies the memory to a new swap file (using copy-on-write if available)
+    #[cfg(not(target_os = "windows"))]
+    pub fn fork(&mut self, hint_used: Option<usize>) -> Result<Self, String> {
+        // Empty memory is an edge case
+        if self.len == 0 {
+            return Ok(Self::new());
+        }
+
+        // First we sync all the data to the backing file
+        unsafe {
+            libc::fsync(self.fd.0);
+        }
+
+        // Open a new temporary file (which is used for swapping for the forked memory)
+        let fd = unsafe {
+            let file = libc::tmpfile();
+            if file.is_null() {
+                return Err(format!(
+                    "failed to create temporary file - {}",
+                    io::Error::last_os_error()
+                ));
+            }
+            FdGuard(libc::fileno(file))
+        };
+
+        // Attempt to do a shallow copy (needs a backing file system that supports it)
+        unsafe {
+            if libc::ioctl(fd.0, 0x94, 9, self.fd.0) != 0
+            // FICLONE
+            {
+                #[cfg(feature = "tracing")]
+                trace!("memory copy started");
+
+                // Determine host much to copy
+                let len = match hint_used {
+                    Some(a) => a,
+                    None => self.len,
+                };
+
+                // The shallow copy failed so we have to do it the hard way
+                let mut off_in: libc::off_t = 0;
+                let mut off_out: libc::off_t = 0;
+
+                cfg_if::cfg_if! {
+                    if #[cfg(not(any(target_env = "musl", target_vendor = "apple")))]
+                    {
+                        let ret = libc::copy_file_range(self.fd.0, &mut off_in, fd.0, &mut off_out, len, 0);
```

Comment:
This is causing the CI failure on linux-x64:

```
ld.lld --error-limit=0 -O3 -z
stack-size=16777216 --gc-sections --eh-frame-hdr -znow -m
aarch64linux -o /media/psf/Home/Development/wasmer/tests/integration/cli/debug/python
/home/parallels/.cache/zig/o/e87e9a86d6703349199f944fa8e74a29/Scrt1.o
/home/parallels/.cache/zig/o/e7ccbcf4c6260284654d7e852778c92c/crti.o -dynamic-linker
/lib/ld-linux-aarch64.so.1 /media/psf/Home/Development/wasmer/tests/integration/cli/debug/python.o
/media/psf/Home/Development/wasmer/tests/integration/cli/debug/libwasmer.a
/media/psf/Home/Development/wasmer/tests/integration/cli/debug/volume.obj
/home/parallels/.cache/zig/o/b0b7a26d74bbb6c7060ad372b02cbda0/wasmer_main.o
--as-needed /home/parallels/.cache/zig/o/f9abe4dcff02854b182b64b8691cfec5/libunwind.a
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libm.so.6
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libpthread.so.0
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libc.so.6
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libdl.so.2
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/librt.so.1
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libld.so.2
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libutil.so.1
/home/parallels/.cache/zig/o/59a82fa70408f42b72de3e9040755e78/libresolv.so.2
/home/parallels/.cache/zig/o/1952c727694664ba6c3c698fa8822f61/libc_nonshared.a
/home/parallels/.cache/zig/o/0acdc9e72b88424c9e3fd3a49be30b1a/crtn.o --allow-shlib-undefined
ld.lld: error: undefined symbol: copy_file_range
>>> referenced by wasmer_vm.bc74d6e7-cgu.3
>>>               wasmer_vm-992d0331437bc80b.wasmer_vm.bc74d6e7-cgu.3.rcgu.o:(wasmer_vm::mmap::Mmap::fork::h7e88e1f027f5f888)
                  in
                  archive /media/psf/Home/Development/wasmer/tests/integration/cli/debug/libwasmer.a
                  Error: python_exe_output_path debug/python does not
                  exist thread 'test_wasmer_create_exe_pirita_works'
                  panicked at 'assertion failed: `(left == right)`
```

However, I'm not sure what can be done about this. An easy fix would be to disable it on linux completely.

```suggestion
                    if #[cfg(not(any(target_family = "linux", target_vendor = "apple")))]
                    {
                        let ret = libc::copy_file_range(self.fd.0, &mut off_in, fd.0, &mut off_out, len, 0);
```

Reply:
Probably this has to do with zigs version of `libc` not being up-to-date.

I'd disable this for now since we have to test the fallback anyway.

Reply:
I've verified that this would fix the CI issue on Linux.

Reply:
I re-wrote the function and started using it on all unix platforms, so we should be good here.

Will re-open if the tests still have problems.